File: LeetCode/Study Plan/Practice Problems/Medium/Minimum Time To Complete Trips/Go/min_time_to_complete_trips.go
package main

import (
	"fmt"
	"sort"
)

func min_time(time []int, total_trips int) int64 {
	sort.Ints(time)
	n := len(time)
	max_time := time[n-1] * total_trips
	low, high := 0, max_time
	for low < high {
		mid_time := (low + high) / 2

		sum := 0
		for i := 0; i < n; i++ {
			sum += (mid_time / time[i])
		}
		if sum >= total_trips {
			high = mid_time
		} else {
			low = mid_time + 1
		}
	}
	return int64(high)
}

func main() {
	fmt.Println(min_time([]int{1, 2, 3}, 5))
}

File: LeetCode/Study Plan/Practice Problems/Medium/Minimum Time To Complete Trips/Java/minTimeToCompleteTrips.java
class Solution {
    public long minimumTime(int[] time, int totalTrips) {
        Arrays.sort(time);

        int size = time.length;
        long l = 1;
        long r = (long)time[size-1]*totalTrips;
        long res = 0;

        while(l <= r) {
            long mid = l+(r-l)/2;
            if (isPossible(time, mid, totalTrips)) {
                res = mid;
                r = mid-1;
            }else {
                l = mid+1;
            }
        }

        return res;
    }

    public boolean isPossible(int[] time, long t, int totalTrips) {
        long trip = 0;

        for (int i = 0; i<time.length; i++) {
            trip += t/(long)time[i];
        }

        return trip >= totalTrips;
    }
}
